fuel-asm/derive/src/codegen.rs

```diff
@@ -0,0 +1,507 @@
+            pub const OPCODE: Opcode = Opcode::#opcode_name;
+        }
+    })
+}
+
+pub fn op_debug_impl(instructions: &InstructionList) -> TokenStream {
+    instructions.map_to_tokens(|Instruction {
+        opcode_name,
+        args,
+        ..
+    }| {
+        let values: TokenStream = comma_separated(args.names());
+        let fields: TokenStream = args.map_to_tokens(|arg| {
+            let name = &arg.name;
+            if arg.is_imm() {
+                quote! {
+                    .field(stringify!(#name), &format_args!("{}", #name.to_smallest_int()))
+                }
+            } else {
+                quote! {
+                    .field(stringify!(#name), &format_args!("{:#02x}", u8::from(#name)))
+                }
+            }
+        });
+
+        let unpack_if_needed = if args.is_empty() {
+            quote! {}
+        } else {
+            quote! {
+                let (#values) = self.unpack();
+            }
+        };
+
+        quote! {
+            impl core::fmt::Debug for #opcode_name {
+                #[warn(clippy::unused_unit)] // Simplify code
+                fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+                    #unpack_if_needed
+                    f.debug_struct(stringify!(#opcode_name))
+                        #fields
+                        .finish()
+                }
+            }
+        }
+    })
+}
+
+pub fn opcode_enum(instructions: &InstructionList) -> TokenStream {
+    let variants: TokenStream = instructions.map_to_tokens(
+        |Instruction {
+             description,
+             opcode_name,
+             opcode_number,
+             ..
+         }| {
+            quote! {
+                #[doc = #description]
+                #opcode_name = #opcode_number,
+            }
+        },
+    );
+    quote! {
+        #[doc = "The opcode numbers for each instruction."]
+        #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+        #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+        #[cfg_attr(feature = "typescript", wasm_bindgen::prelude::wasm_bindgen)]
+        pub enum Opcode {
+            #variants
+        }
+    }
+}
+
+pub fn opcode_try_from(instructions: &InstructionList) -> TokenStream {
+    let arms = instructions.map_to_tokens(
+        |Instruction {
+             opcode_number,
+             opcode_name,
+             ..
+         }| {
+            quote! {
+                #opcode_number => Ok(Opcode::#opcode_name),
+            }
+        },
+    );
+    quote! {
+        impl core::convert::TryFrom<u8> for Opcode {
+            type Error = InvalidOpcode;
+
+            fn try_from(value: u8) -> Result<Self, Self::Error> {
+                match value {
+                    #arms
+                    _ => Err(InvalidOpcode),
+                }
+            }
+        }
+    }
+}
+
+pub fn from_op(instructions: &InstructionList) -> TokenStream {
```

Comment:
I like how most other codegen functions are named with the type they are implementing the behavior on first, e.g `op_try_from`, `op_debug_impl`, `instruction_try_from_bytes`. Therefore it would be nice to name this function in such a way that it leads with `op`.

I understand that `op_from` may sound like conversion methods that create opcodes from other types, so perhaps `op_into` would be a better name. Or just `op_conversions`.

I'm okay leaving this as-is as well. What are your thoughts?

Reply:
That's a good point. Renamed to `op_conversions` in 43172b1c